bin/ocamllsp.ml

```diff
@@ -0,0 +1,133 @@
+          "the current dune project lacks a lockdir associated with the default context"
+      | No_ocaml_lockfile_in_lock_dir { lock_dir_path } ->
+        sprintf
+          "the lockdir %S lacks a lockfile for the %S package"
+          (Path.Source.to_string lock_dir_path)
+          (Package_name.to_string compiler_package_name)
+    ;;
+  end
+
+  (* A message explaining why dune can't install/run ocamllsp as a
+     dev-tool. *)
+  let message reason =
+    User_message.make
+      [ Pp.textf
+          "Unable to run %s as a dev-tool because %s."
+          ocamllsp_exe_name
+          (Reason.to_string reason)
+      ; Pp.nop
+      ; Pp.textf
+          "To run %s as a dev-tool, the following conditions mest be met:"
+          ocamllsp_exe_name
+      ; Pp.text " - You must be inside a dune project."
+      ; Pp.text " - The dune project must have a lockdir for its default context."
+      ; Pp.textf
+          " - The lockdir must contain a lockfile for the package %S."
+          (Package_name.to_string compiler_package_name)
+      ; Pp.nop
+      ; Pp.textf
+          "This is because %s must be compiled with the same version of the ocaml \
+           compiler as the project it's analyzing, and without a lockfile for the ocaml \
+           compiler package (ie. %S) the appropriate version of the compiler to use to \
+           compile %s is not known."
+          ocamllsp_exe_name
+          (Package_name.to_string compiler_package_name)
+          ocamllsp_exe_name
+      ]
+  ;;
+end
+
+(* Replace the current dune process with ocamllsp. *)
+let run_ocamllsp common ~args =
+  let exe_path_string = Path.to_string ocamllsp_exe_path in
+  Console.print_user_message
+    (Dune_rules.Pkg_build_progress.format_user_message
+       ~verb:"Running"
+       ~object_:
+         (User_message.command (String.concat ~sep:" " (ocamllsp_exe_name :: args))));
+  Console.finish ();
+  restore_cwd_and_execve common exe_path_string (exe_path_string :: args) Env.initial
+;;
+
+let build_ocamllsp common =
+  let open Fiber.O in
+  let+ result =
+    Build_cmd.run_build_system ~common ~request:(fun _build_system ->
+      Action_builder.path ocamllsp_exe_path)
+  in
+  match result with
+  | Error `Already_reported -> raise Dune_util.Report_error.Already_reported
+  | Ok () -> ()
+;;
+
+(* Ocamllsp needs to be compiled with the same version of the ocaml
+   compiler as the code it's analyzing, so look up the version of the
+   ocaml compiler in the lockdir so it can be added to the constraints
+   when solving ocamllsp's lockdir. *)
+let locked_ocaml_compiler_version () =
+  let open Memo.O in
+  let context =
+    (* Dev tools are only ever built with the default context. *)
+    Context_name.default
+  in
+  let* result = Dune_rules.Lock_dir.get context in
+  match result with
+  | Error _ -> Memo.return @@ Error Failure.Reason.No_lock_dir
+  | Ok { packages; _ } ->
+    (match Package_name.Map.find packages compiler_package_name with
```

Comment:
The way this works is by matching the version number of OCaml inside the project with the version number we're going to solve for. While this will usually work, it will fail for custom versions of the compiler.

The more correct thing to do would be to "pin" the compiler in LSP to be the compiler from the project. Don't think this is strictly necessary here, but we should leave a comment + add a test case for this situation.

Reply:
I added a comment explaining this.